Event Log entry when node joins cluster

mrtracy opened the pull request:
Nodes will now log an event into the `system.eventlog` table after joining a
cluster. The event type is different depending on if this is the first time a
node is joining the cluster: if it is the first time, a "node_join" event is
logged. Otherwise, a "node_restart" event is logged.

Includes a new acceptance test to verify that this event is logged correctly in
the multi-node and restart case.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5550)

<!-- Reviewable:end -->

Commit:
Event Log entry when node joins cluster

Nodes will now log an event into the `system.eventlog` table after joining a
cluster. The event type is different depending on if this is the first time a
node is joining the cluster: if it is the first time, a "node_join" event is
logged. Otherwise, a "node_restart" event is logged.

Includes a new acceptance test to verify that this event is logged correctly in
the multi-node and restart case.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[acceptance/event_log_test.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKfMBJ1SDChBQitxF:-KDeKfMBJ1SDChBQitxG:-1719044135) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L26)):*
universally called `csql` elsewhere

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDeKjII_OBC9Gf5Qas8:-1315989731) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
is it time to move this into `runTestOnConfigs`?

---

*[acceptance/event_log_test.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfC-CWpTKbPAiubZh7:-KDfC-CX0ZNsk2a3w6Ff:-1619154022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L71)):*
probably should straight up t.Fatal here

---

*[acceptance/event_log_test.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfC0xwnc8F--iQaH4C:-KDfC0xwnc8F--iQaH4D:-1619154022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L80)):*
probably should straight up t.Fatal here

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDfCuL8u6vDNlcRRWmJ:-1619154022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
probably should straight up t.Fatal here

---

*[acceptance/event_log_test.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCteUeJ4zshLm3Rjw:-KDfCteUeJ4zshLm3Rjx:-1619154022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L141)):*
probably should straight up t.Fatal here

---

*[acceptance/event_log_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCwJC7VqEklbMrqEh:-KDfCwJC7VqEklbMrqEi:-1619154022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L150)):*
probably should straight up t.Fatal here

---

*[server/node.go, line 635 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeIL0b_HZ0txOYyl8X:-KDeIL0b_HZ0txOYyl8Y:1644183948) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/server/node.go#L635)):*
`complete` seems superfluous.

```go
	logEventType := sql.EventLogNodeRestart
	if n.initialBoot {
		logEventType = sql.EventLogNodeJoin
	}

	n.stopper.RunWorker(func() {
		for {
			if err := n.ctx.DB.Txn(func(txn *client.Txn) *roachpb.Error {
				return sql.MakeEventLogger(n.ctx.SQLExecutor.LeaseManager).InsertEventRecord(txn,
					logEventType,
					int32(n.Descriptor.NodeID),
					int32(n.Descriptor.NodeID),
					struct {
						Descriptor roachpb.NodeDescriptor
						ClusterID  uuid.UUID
						StartedAt  int64
					}{n.Descriptor, n.ClusterID, n.startedAt},
				)
			}); err != nil {
				log.Warningc(n.context(), "unable to log %s event for node %d: %s", logEventType, n.Descriptor.NodeID, err)
			} else {
				return
			}

			select {
			case <-n.stopper.ShouldStop():
				return
			case <-time.After(time.Second):
				// Continue
			}
		}
	})
```

---

*[server/node.go, line 660 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeJJvKFU6UA1-Qozkl:-KDeJJvLMwsd5lkXOl6_:1149608200) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/server/node.go#L660)):*
do you need this backoff? if not, better to not allocate a channel on each iteration. if yes, you should allocate the timer once and reuse it. see util.Timer.

---

*[util/uuid/uuid.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeI67Qekjpn4ZcT-nT:-KDeI67RaBtQOkHFFWLk:862128341) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/util/uuid/uuid.go#L87)):*
other comments here are of the form "... delegates to ..." - any reason to deviate?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-:-KDfCz0s5OqMZN2Fcvhr:1417088721)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[acceptance/event_log_test.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKfMBJ1SDChBQitxF:-KDg2KfNUtOIQk__PsOW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L26)):*
Done.

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDg2Mcfyt3PQu_p-aYq:-1677549562) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
Now that, I don't know. Not an expert on these here acceptance tests.

---

*[acceptance/event_log_test.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfC-CWpTKbPAiubZh7:-KDg2gzezsJTugMs2U9K:-1922466769) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L71)):*
Yes, there's no recovering from this. Done.

---

*[acceptance/event_log_test.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfC0xwnc8F--iQaH4C:-KDg2o3BD-CqmfwbawBb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L80)):*
Done.

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDg2ur7kkb3Uly65h2Q:-1117847908) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
This error can be transient.

---

*[acceptance/event_log_test.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCteUeJ4zshLm3Rjw:-KDg2z72TkBX7YgAbD4y:1528239032) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L141)):*
It's like you're commenting on pull requests with multiple cursors. Done.

---

*[acceptance/event_log_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCwJC7VqEklbMrqEh:-KDg3BQpuZGvopblX9FL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L150)):*
Done.

---

*[server/node.go, line 635 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeIL0b_HZ0txOYyl8X:-KDg4YypfPh2ODzRH7zX:2040445016) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/server/node.go#L635)):*
Yes, that is nicer. Done.

---

*[server/node.go, line 660 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeJJvKFU6UA1-Qozkl:-KDg4_2wm8q6qGx_xUOC:-415720524) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/server/node.go#L660)):*
Eh, I probably don't need the backoff.  Replaced with `default:` case.

---

*[util/uuid/uuid.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeI67Qekjpn4ZcT-nT:-KDg7jqRJ_Aod2J3up7m:-1811755399) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/util/uuid/uuid.go#L87)):*
Just didn't notice. Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDhR8hUAe7nca-c-QOB:1931032663) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
cc @bramgruneir 

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDhQqBc8fZLWuHKnXqB:-865489721) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
Really? What is the transient error here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDil3T03uz6dGE5BTdW:-1633600699) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
It should.  And this looks good.  Right now running them against the list of configs is still a manual process, via the flag `-test-configs`.  I'll be enabling the flag in circle when I think we're ready.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDj0UzDg5bsV6PTCvVe:1172202722) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
`event_log_test.go:133: dial tcp 192.168.99.100:32856: getsockopt: connection refused`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDj3gmi0Uj3z9wtfcSl:-674922433) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
@mrtracy did you mean to do this?

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDj3p5yRBZKVuvaJ4mT:2077881636) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
Oh, because the node is restarting. I think this needs a comment, since there's a very similar query above `if err := rows.Scan(&targetID, &infoStr); err != nil {` that does actually t.Fatal on error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[acceptance/event_log_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDeKjII_OBC9Gf5Qas7:-KDjQwh-3cKyWDVZLIj7:-1410218496) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L38)):*
No i did not, it would be easy as pie for anyone to do in a separate PR.

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDjRChsQS-V_H8aj7IJ:-1644456356) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
I do not believe that `rows.Scan()` can be transient; at that point, `rows.Next()` has succeeded and a failure implies the structure of the row is incorrect.  I think that the errors returned from db.Query() are the only ones subject to network errors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[acceptance/event_log_test.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550#-KDfCuL8u6vDNlcRRWmI:-KDjUcgT7vJs9CFhqjCp:-446347396) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ab1a0b8e522b430246e09b9aae2abc31f436e84/acceptance/event_log_test.go#L133)):*
Ah, you're right - you're treating `rows.Err()` as retryable. My mistake.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5550)*
<!-- Sent from Reviewable.io -->
